### PR TITLE
Handle Unlocking of Waypoints and Statues

### PR DESCRIPTION
**Pull request by GanyusLeftHorn**
## Description

This PR adds the ability to unlock both Statues of the Seven as well as teleport waypoints (with the exception of the very first statue, which seems to be tied to another questline). The work done in this PR includes:
- Merging management of open states and map unlocks into a new `PlayerProgressManager`.
- Automatically start subquests of the `303` questline, which is necessary to make statues interactable.
- Fix some issues with our quests implementation to make unlock triggers for statues work.
- Handle unlocking of both statues (and their corresponding map areas) and waypoints, including the very generous reward of five primogems per unlock. :P

The first statue is added to the list of unlocked areas and waypoint automatically, in order to not make it permanently locked off.

## Type of changes

- [ ] Bug fix
- [x] New feature 
- [ ] Enhancement
- [ ] Documentation

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] My pull request is unique and no other pull requests have been opened for these changes
- [x] I have read the [Contributing note](https://github.com/Grasscutters/Grasscutter/blob/stable/CONTRIBUTING.md) and [Code of conduct](https://github.com/Grasscutters/Grasscutter/blob/development/CODE_OF_CONDUCT.md)
- [x] I am responsible for any copyright issues with my code if it occurs in the future.

### CODE REVIEW

**Comment by millet6** [comment]
太好了 有了这个 Great, with this

File: src/main/java/emu/grasscutter/server/packet/recv/HandlerUnlockTransPointReq.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+import emu.grasscutter.net.packet.PacketHandler;
+import emu.grasscutter.server.game.GameSession;
+
+@Opcodes(PacketOpcodes.UnlockTransPointReq)
+public class HandlerUnlockTransPointReq extends PacketHandler {
+    @Override
+    public void handle(GameSession session, byte[] header, byte[] payload) throws Exception {
+        UnlockTransPointReq req = UnlockTransPointReq.parseFrom(payload);
+        session.getPlayer().getProgressManager().unlockTransPoint(req.getSceneId(), req.getPointId());
+    }
```
Comment (akatatsu27) [inline src/main/java/emu/grasscutter/server/packet/recv/HandlerUnlockTransPointReq.java:15]: I also got PlayerPropChangeReasonNotify, PlayerPropNotify, and ItemAddHintNotify for adventure exp. official server sent me 10, but you get 50 adventure exp, so maybe the exp modifier should be changed? There's also the handbook investigation target, but that is not implemented yet
```suggestion
    }
    session.getPlayer().getQuestManager().triggerEvent(QuestTrigger.QUEST_CONTENT_UNLOCK_TRANS_POINT, req.getSceneId(), req.getPointId());

        // Give Adventure Exp.
        session.getPlayer().getInventory().addItem(102,10, ActionReason.getTypeByValue(19));
        session.getPlayer().earnExp(10);

        session.send(new PacketPlayerPropChangeReasonNotify(session.getPlayer().getProperty(PROP_PLAYER_EXP),
            PROP_PLAYER_EXP, PropChangeReasonOuterClass.PropChangeReason.PROP_CHANGE_REASON_PLAYER_ADD_EXP));
```

File: src/main/java/emu/grasscutter/game/player/PlayerProgressManager.java [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+        this.tryUnlockOpenStates(false);
+
+        // Send notify to the client.
+        player.getSession().send(new PacketOpenStateUpdateNotify(this.player));
+
+        // Add statue quests if necessary.
+        this.addStatueQuestsOnLogin();
+
+        // Auto-unlock the first statue and map area, until we figure out how to make
+        // that particular statue interactable.
```
Comment (akatatsu27) [inline src/main/java/emu/grasscutter/game/player/PlayerProgressManager.java:46]: The first statue is unlocked via the mainQuest 352. The roadblock for this, and a lot of early and important quests is that luaj is too stupid for `QUEST_CONTENT_TRIGGER_FIRE`, and it won't work, so quest 352 can't continue to the statue unlock cutscene. I have made contentTriggerFire work using tnsc4502's jnlua in this [branch "questing-jnlua" in my repository](https://github.com/akatatsu27/Grasscutter/tree/questing-jnlua) , with no issues, and minimal changes. By simply switching to jnlua we can actually start working on the early game properly.

Comment (GanyusLeftHorn) [reply src/main/java/emu/grasscutter/game/player/PlayerProgressManager.java:46]: Yes, I know it _should_ be tied to 352, but I couldn't get the statue to reliably unlock even when I manually finished that questline via commands. Probably an issue in the quest manager, but I didn't want to look into that too much. I even managed to brick my account when trying various combinations of subquests, with the quest manager throwing exceptions on login. :P

File: src/main/java/emu/grasscutter/server/packet/recv/HandlerUnlockTransPointReq.java (hunk 1 above)
Comment (GanyusLeftHorn) [reply src/main/java/emu/grasscutter/server/packet/recv/HandlerUnlockTransPointReq.java:15]: Oh, I completely forgot that unlocking waypoints gives Adventure EXP. I'll add that tomorrow.

**Review by ghost — APPROVED** [review]

**Review by Birdulon — APPROVED** [review]

**Comment by Birdulon** [comment]
Very cool

**Comment by eternalcomet** [comment]
can you add the feature of worshiping the statue?能否添加供奉神像功能？

**Comment by longyi8** [comment]
Can all anchors be unlocked by command?
能否通过指令解锁全部的锚点？

**Comment by eternalcomet** [comment]
> Can all anchors be unlocked by command? 能否通过指令解锁全部的锚点？

perhaps you can use command `/setprop unlockmap`
好像可以通过命令/setprop unlockmap解锁

**Comment by zxp1314gch** [comment]
> 
invalid command
无效指令

**Comment by ZxyGch** [comment]
> > Can all anchors be unlocked by command? 能否通过指令解锁全部的锚点？
> 
> perhaps you can use command `/setprop unlockmap` 好像可以通过命令/setprop unlockmap解锁

invalid command
无效指令

**Comment by longyi8** [comment]
> > > Can all anchors be unlocked by command? 能否通过指令解锁全部的锚点？
> > 
> > 
> > perhaps you can use command `/setprop unlockmap` 好像可以通过命令/setprop unlockmap解锁
> 
> invalid command 无效指令

I can use /prop unlockmap 1
我可以使用/prop unlockmap 1

**Comment by ZxyGch** [comment]
> I can use /prop unlockmap 1 我可以使用/prop unlockmap 1
thank you very much！
非常感谢！

**Pull request merged**